concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1147 @@
+	subscriptionWithoutWebhookAuth,
+	WithPath(),
+	WithServiceAfter(svcName, host), // instead pass it as an argument to NewAPIRule
+)
+```
+</details>
+
+**Example 3**: Side effect as new function
+
+<details>
+  <summary>Don't</summary>
+
+`WithStatusReady` is not used inside a function that accepts a functional option as input.
+As a reader, you would not expect that it performs a **side-effect** on the `apiRuleNew`.
+
+```go
+// source: https://github.com/kyma-project/kyma/blob/a84f76f674babfc63e369bbba44de76499cf475d/components/eventing-controller/controllers/subscription/beb/reconciler_test.go#L592
+getAPIRule(ctx, apiRuleNew).Should(And(
+	HaveNotEmptyHost(),
+	HaveNotEmptyAPIRule(),
+))
+WithStatusReady(apiRuleNew) // <= WithStatusReady not used inside a function that accepts a functional option as input (for example, `APIRuleOption`)
+```
+
+</details>
+
+<details>
+  <summary>Do</summary>
+
+To make the side-effect more obvious, the following example uses a function named `MarkReady` instead.
+
+```go
+getAPIRule(ctx, apiRuleNew).Should(And(
+	HaveNotEmptyHost(),
+	HaveNotEmptyAPIRule(),
+))
+MarkReady(apiRuleNew) // instead consider using another function that explicitly states from the name that it applies a side effect
+```
+</details>
+
+## Testing pyramid in Kyma
+
+The following section describes the different testing levels for Kyma. These levels are usually seen as a [pyramid](https://en.wikipedia.org/wiki/Test_automation#Testing_at_different_levels):
+- At the bottom of the pyramid, you should have the **most** test cases because they are usually very **fast**. However, you are limited in what you can test because external dependencies are **mocked**. Unit tests are an example of this.
+- At the top of the pyramid, there are usually **fewer** tests because these tests take **longer**. However, these tests are a more realistic scenario of how the end user will use our product. E2E tests are an example of this.
+
+### Controller Integration Tests (kubebuilder)
+
+**Test Setup**
+Controllers using the kubebuilder framework bootstrap integration tests using `Ginkgo` (for testing) and `Gomega` (for assertions).
+Using the `controller-runtime/pkg/envtest` package, a kubernetes **control-plane** (**API server** and **etcd**) is started locally once before the first test runs (`BeforeSuite`).
+You can also use a custom Kubernetes cluster (see [USE_EXISTING_CLUSTER](https://book.kubebuilder.io/reference/envtest.html) environment variable).
+
+**Goal**
+The goal of a controller integration test is to test the Kubernetes controller in a **limited** Kubernetes **environment**. That means that dependencies of the controller - such as other controllers (APIRule controller) or external eventing system - are not present. These systems are only present in higher hierarchies of the testing pyramid.
+
+Controller integration tests should be reserved for cases when we care about the eventual state of the object or need a Kubernetes cluster.
```

Comment:
Do you mean "eventual" as in "final", or is this a special meaning related to "Eventing" that I'm not familiar with? If "final", I'd suggest writing "final" :D

```suggestion
Controller integration tests should be reserved for cases when we care about the final state of the object or need a Kubernetes cluster.
```

Reply:
I think eventual is the better fit (coming from **eventual consistency** in kubernetes).

[source](https://downey.io/blog/desired-state-vs-actual-state-in-kubernetes/#:~:text=Since%20Kubernetes%20is%20optimized%20for,eventually%2C%20the%20system%20will%20converge):
> Since Kubernetes is optimized for running resilient, highly available workloads, it chooses Availability. State updates are propagated over time in what is known as [eventual consistency](https://en.wikipedia.org/wiki/Eventual_consistency). Eventual consistency means that if no new updates are made to a system, then eventually, the system will converge

Example from [kubernetes blog](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/): 
> The design aim is for you to be able to request deletion and know when processes terminate, but also be able to ensure that deletes eventually complete.